docs/proposals/ssh-known_hosts.rst

```diff
@@ -0,0 +1,93 @@
+tunnelling and proxy commands for other services, including, connecting to the
+Vault cluster and the private Kubernetes API server endpoint. Currently in
+development is the replacement of our programmatic use cases of SSH in favour of
+the in package Go solution, a choice stemming from pain points developing more
+sophisticated utility functions for Tarmak and the desire for improvements in
+control of connections to remote hosts.
+
+During development of this replacement it became clear that proper care must be
+taken during authentication of host public keys during connection and manual
+management of our ``ssh_known_hosts`` cluster file. Our current implementation
+allows OpenSSH to maintain this file however, does not exit with an error if
+public keys do not match due to the flag ``StrictHostKeyChecking`` set to
+``no``. Not only does a miss-match in public keys not cause an error, the
+population of known public keys on different authenticated machines to the same
+EC2 hosts will always use the hosts presented public key, meaning the set of
+public keys could potentially be different for users accessing the same cluster.
+
+Objective
+---------
+
+By implementing stricter enforcement of the ``ssh_known_hosts`` file and passing
+it's management to Tarmak, we can improve the security of SSH connections to
+remote hosts. The key high level points to achieving this is as follows:
+
+ - Disable writes from the OpenSSH command to the ``ssh_known_hosts`` file and
+   enforce strict checking.
+ - Enforce that our in package implementation of SSH connections adheres to this
+   file also.
+ - Collect public keys during instance start up that are then stored, tightly
+   coupled with that host. These keys are able to be used as a source of truth
+   for other authenticated users attempting to connect to remote hosts on the
+   cluster that have empty or an incomplete ``ssh_known_hosts`` file.
+
+Changes
+-------
+
+Firstly, we must restrict the OpenSSH command line tool from editing the
+``ssh_known_hosts`` file and strictly enforce it by updating the generator for the
+``ssh_config`` file. This enables Tarmak to take control of the ``ssh_known_hosts``
+file management.
+
+In order to create a source of truth for each host's public key, each instance
+will have it's public key attached as a tag, shortly after boot time like the
+following:
+
++------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------+
+| public_key | ecdsa-sha2-nistp256 AAAAE2VjZHNhLXNoYTItbmlzdHAyNTYAAAAIbmlzdHAyNTYAAABBBPF+xkIGMUNVI0gElRaTLjfA4QMN/XGJhHswDyv59DNSOtG3KwZvDF3YkAb0PkTQAYo8N5fxoKqimGugOAaefPc= |
++------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------+
+
+The population of this tag will occur during a Terraform apply in which the
```

Comment:
Doesn't tarmak also use 'autoscaling groups' to autoprovision instances? How will these have the tag populated?

Reply:
This is the tricky bit here. In theory the nodes could self update it's tags. But I think the AWS IAM Policies can't be configured that tight to allow only self update. This also has some issues, once the node is overtaken and other tags (e.g. tarmak_role) are no longer safe.

Another problem is that the amount of tags to be stored on an AWS instance cannot exceed 255 chars. Also we have more than just the ecdsa host key.

Reply:
I also think we have to look further than AWS. Not every provider will have support for tags. I think we should come up with a universal solution for this.

Reply:
While I agree, IMO taging instances on AWS would be the best way to couple the public keys with the instance life cycle. If we keep the SSH config with,
`HostKeyCallback: ssh.tarmak.Provider().HostKeyCallBack,`
then each provider can have it's specific implementation. 
This is assuming all targeted providers will have some kind of instance metasdata we can pull and push from